source/MRMesh/MRMeshLoadObj.cpp

```diff
@@ -143,141 +146,158 @@ tl::expected<std::vector<NamedMesh>, std::string> fromSceneObjFile( const char*
+            taskGroup.run( [&, i = gi]
+            {
+                auto& group = groups[i];
+                const auto begin = i * groupSize;
+                const auto end = std::min( ( i + 1 ) * groupSize, size );
+                for ( auto ci = begin; ci < end; ci++ )
+                    if ( data[ci] == '\n' )
+                        group.emplace_back( ci + 1 );
+            } );
+        }
+        taskGroup.wait();
+
+        size_t sum = 1;
+        for ( const auto& group : groups )
+            sum += group.size();
+        newlines.reserve( sum );
+        for ( const auto& group : groups )
+        {
+            assert( newlines.back() < group.front() );
+            newlines.insert( newlines.end(), group.begin(), group.end() );
         }
-    }, tbb::static_partitioner() );
-    std::vector<std::vector<size_t>> newlinesBlocks;
-    size_t newlinesSize = 1;
-    for ( auto&& newlinesBlock : newlinesPerThread )
-    {
-        assert( !newlinesBlock.empty() );
-        newlinesSize += newlinesBlock.size();
-        newlinesBlocks.emplace_back( std::move( newlinesBlock ) );
     }
-    std::sort( newlinesBlocks.begin(), newlinesBlocks.end(), [] ( const auto& lhs, const auto& rhs )
-    {
-        return lhs.front() < rhs.front();
-    } );
-    newlines.reserve( newlinesSize );
-    for ( auto&& newlinesBlock : newlinesBlocks )
-        newlines.insert( newlines.end(), newlinesBlock.begin(), newlinesBlock.end() );
     // add finish line
     if ( newlines.back() != size )
         newlines.emplace_back( size );
+    const auto lineCount = newlines.size() - 1;
 
     if ( !callback( 0.40f ) )
         return tl::make_unexpected( "Loading canceled" );
 
-    timer.restart( "find element lines" );
-    struct Object
+    timer.restart( "group element lines" );
+    enum Element
     {
-        std::string name;
-        std::vector<size_t> faceLines;
+        Unknown,
+        Vertex,
+        Face,
+        Object,
     };
-    std::vector<size_t> vertexLines;
-    std::vector<Object> objects( 1 ); // emplace default object
-    for ( size_t li = 0; li + 1 < newlines.size(); ++li )
+    struct ElementGroup
+    {
+        Element element;
+        size_t begin;
+        size_t end;
+    };
+    std::vector<ElementGroup> groups{ { Unknown, 0, 0 } }; // emplace stub initial group
+    for ( size_t li = 0; li < lineCount; ++li )
     {
         auto* line = data + newlines[li];
+
+        Element element = Unknown;
         if ( line[0] == 'v' && line[1] != 'n' /*normals*/ && line[1] != 't' /*texture coordinates*/ )
         {
-            vertexLines.emplace_back( li );
+            element = Vertex;
         }
         else if ( line[0] == 'f' )
         {
-            objects.back().faceLines.emplace_back( li );
+            element = Face;
         }
         else if ( line[0] == 'o' )
         {
-            auto object = objects.emplace_back();
+            element = Object;
+        }
+        // TODO: multi-line elements
```

Comment:
What is multi-line elements?

Reply:
Example:
```
bmat u  1 -3  3 -1 \
        0  3 -6  3 \
        0  0  3 -3 \
        0  0  0  1
```
We are not using such elements right now, but we may need their support in the future.